staging/src/k8s.io/apiserver/plugin/pkg/audit/dynamic/dynamic.go

```diff
@@ -0,0 +1,306 @@
+	}
+	cm, err := webhook.NewClientManager(auditv1.SchemeGroupVersion, func(s *runtime.Scheme) error {
+		auditinstall.Install(s)
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	authInfoResolver, err := webhook.NewDefaultAuthenticationInfoResolver("")
+	if err != nil {
+		return nil, err
+	}
+	cm.SetAuthenticationInfoResolver(authInfoResolver)
+	cm.SetServiceResolver(c.WebhookConfig.ServiceResolver)
+	cm.SetAuthenticationInfoResolverWrapper(c.WebhookConfig.AuthInfoResolverWrapper)
+
+	manager := &backend{
+		config:               c,
+		delegates:            atomic.Value{},
+		delegateUpdateMutex:  sync.Mutex{},
+		webhookClientManager: cm,
+		recorder:             recorder,
+	}
+	manager.delegates.Store(syncedDelegates{})
+
+	return manager, nil
+}
+
+type backend struct {
+	// delegateUpdateMutex holds an update lock on the delegates
+	delegateUpdateMutex  sync.Mutex
+	config               *Config
+	delegates            atomic.Value
+	webhookClientManager webhook.ClientManager
+	recorder             record.EventRecorder
+}
+
+type syncedDelegates map[types.UID]*delegate
+
+// Names returns the names of the delegate configurations
+func (s syncedDelegates) Names() []string {
+	names := []string{}
+	for _, delegate := range s {
+		names = append(names, delegate.configuration.Name)
+	}
+	return names
+}
+
+// ProcessEvents proccesses the given events per current delegate map
+func (b *backend) ProcessEvents(events ...*auditinternal.Event) {
+	for _, d := range b.GetDelegates() {
+		d.ProcessEvents(events...)
+	}
+}
+
+// Run starts a goroutine that propagates the shutdown signal,
+// individual delegates are ran as they are created.
+func (b *backend) Run(stopCh <-chan struct{}) error {
+	go func() {
+		<-stopCh
+		b.disableUpdates()
+		b.stopAllDelegates()
+	}()
+	// event handler is added after the stop handler is provisioned, this is to prevent any possible
+	// race condition involving the informer adding sinks before Run is called. This will still happen
+	// before the informers are started
+	b.config.Informer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			b.addSink(obj.(*auditregv1alpha1.AuditSink))
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			b.updateSink(oldObj.(*auditregv1alpha1.AuditSink), newObj.(*auditregv1alpha1.AuditSink))
+		},
+		DeleteFunc: func(obj interface{}) {
+			b.deleteSink(obj.(*auditregv1alpha1.AuditSink))
+		},
+	})
+	return nil
+}
+
+// disableUpdates permanently disables the ability to modify delegates
+// this is used to shutdown gracefully
+func (b *backend) disableUpdates() {
+	b.delegateUpdateMutex.Lock()
+}
+
+// stopAllDelegates closes the stopChan for every delegate to enable
+// goroutines to terminate gracefully
+func (b *backend) stopAllDelegates() {
+	for _, d := range b.GetDelegates() {
+		close(d.stopChan)
+	}
+}
+
+// Shutdown calls the shutdown method on all delegates
+func (b *backend) Shutdown() {
```

Comment:
what is the difference between calling Shutdown() on the backend and closing the `stopCh` passed into Run()? I would only have expected one of those mechanisms to exist

Reply:
so `stopAllDelegates` will cause the buffers to purge out their remaining events, and shutdown will finish shutting things down. This was originally one method but was asked to be split, it needs to be handled slightly differently then the other plugins because backends can be removed arbitrarily.

Reply:
> `stopAllDelegates` will cause the buffers to purge out their remaining events, and shutdown will finish shutting things down

I'm still having trouble understanding when I would call one verses the other (or if both are required to be called in a particular sequence).